File: js/src/whalesong/groupMetadata.js
import {
  CollectionManager,
  ModelManager
} from './common.js';
import {
  command
} from '../manager.js';

export class ParticipantManager extends ModelManager {

  static mapModel(item) {
    return Object.assign(ModelManager.mapModel(item), {
      id: item.id._serialized
    });
  }
}

export class ParticipantCollectionManager extends CollectionManager {

  static getModelManagerClass() {
    return ParticipantManager;
  }

  _getParticipants(contactIds) {
    return contactIds.map((contactId) => this.collection.get(contactId)).filter(Boolean);
  }

  _getContacts(contactIds) {
    return contactIds.map((contactId) => manager.getSubmanager('contacts').collection.get(contactId)).filter(Boolean);
  }

  @command
  async addParticipants({
    contactIds
  }) {
    let contacts = this._getContacts(contactIds);
    return await this.collection.addParticipants(contacts)
  }

  @command
  async canAdd({
    contactId
  }) {
    let contact = manager.getSubmanager('contacts').collection.get(contactId);
    if (!contact) {
      return false;
    }
    return this.collection.canAdd(contacts)
  }

  @command
  async removeParticipants({
    contactIds
  }) {
    let participants = this._getParticipants(contactIds);
    return await this.collection.removeParticipants(participants)
  }

  @command
  async canRemove({
    contactId
  }) {
    let participant = this.collection.get(contactId);
    if (!participant) {
      return false;
    }
    return this.collection.canRemove(participant)
  }

  @command
  async promoteParticipants({
    contactIds
  }) {
    let participants = this._getParticipants(contactIds);
    return await this.collection.promoteParticipants(participants)
  }

  @command
  async canPromote({
    contactId
  }) {
    let participant = this.collection.get(contactId);
    if (!participant) {
      return false;
    }
    return this.collection.canPromote(participant)
  }

  @command
  async demoteParticipants({
    contactIds
  }) {
    let participants = this._getParticipants(contactIds);
    return await this.collection.demoteParticipants(participants)
  }

  @command
  async canDemote({
    contactId
  }) {
    let participant = this.collection.get(contactId);
    if (!participant) {
      return false;
    }
    return this.collection.canDemote(participant)
  }
}


export class GroupMetadataManager extends ModelManager {

  static mapModel(item) {
    return Object.assign(ModelManager.mapModel(item), {
      groupInviteLink: item.groupInviteLink,
      inviteCode: item.inviteCode
    });
  }

  constructor(model) {
    super(model);
    this.addSubmanager('participants', new ParticipantCollectionManager(model.participants));
  }

  @command
  async groupInviteCode() {
    return await this.model.groupInviteCode();
  }

  @command
  async revokeGroupInvite() {
    return await this.model.revokeGroupInvite();
  }

}

export class GroupMetadataCollectionManager extends CollectionManager {

  static getModelManagerClass() {
    return GroupMetadataManager;
  }
}

File: whalesong/managers/group_metadata.py
from typing import List

from dirty_models import ArrayField, BooleanField, DateTimeField, ModelField, StringField, StringIdField

from . import BaseCollectionManager, BaseModelManager
from ..models import BaseModel
from ..results import Result


class Participant(BaseModel):
    is_admin = BooleanField(default=False)
    """
    Whether the participant is a group administrator or not.
    """

    is_super_admin = BooleanField()
    """
    Whether the participant is a group super administrator or not. ¿?
    """


class GroupMetadata(BaseModel):
    """
    Group metadata model.
    """

    announce = StringIdField()
    """
    ¿?
    """

    creation = DateTimeField()
    """
    Group creation timestamp.
    """

    desc = StringField()
    """
    Group description.
    """

    desc_owner = StringIdField()
    """
    Who change last group description.
    """

    desc_time = DateTimeField()
    """
    When last group description was changed.
    """

    owner = StringIdField()
    """
    Who made group.
    """

    participants = ArrayField(field_type=ModelField(model_class=Participant))
    """
    List of participants.
    """

    restrict = StringIdField()
    """
    ¿?
    """

    group_invite_link = StringIdField()
    """
    Group link to invite people.
    """

    invite_code = StringIdField()
    """
    Group code to invite people.
    """


class ParticipantManager(BaseModelManager[Participant]):
    """
    Participant manager.
    """

    MODEL_CLASS = Participant


class ParticipantCollectionManager(BaseCollectionManager[ParticipantManager]):
    """
    Participant collection manager. It allows manage group participants.
    """

    MODEL_MANAGER_CLASS = ParticipantManager

    def add_participants(self, contact_ids: List[str]) -> Result[None]:
        return self._execute_command('addParticipants', {'contactIds': contact_ids})

    def can_add(self, contact_id: str) -> Result[bool]:
        return self._execute_command('canAdd', {'contactId': contact_id})

    def remove_participants(self, contact_ids: List[str]) -> Result[None]:
        return self._execute_command('removeParticipants', {'contactIds': contact_ids})

    def can_remove(self, contact_id: str) -> Result[bool]:
        return self._execute_command('canRemove', {'contactId': contact_id})

    def promote_participants(self, contact_ids: List[str]) -> Result[None]:
        return self._execute_command('promoteParticipants', {'contactIds': contact_ids})

    def can_promote(self, contact_id: str) -> Result[bool]:
        return self._execute_command('canPromote', {'contactId': contact_id})

    def demote_participants(self, contact_ids: List[str]) -> Result[None]:
        return self._execute_command('demoteParticipants', {'contactIds': contact_ids})

    def can_demote(self, contact_id: str) -> Result[bool]:
        return self._execute_command('canDemote', {'contactId': contact_id})


class GroupMetadataManager(BaseModelManager[GroupMetadata]):
    """
    Group metadata manager. It allows manage groups, further than a chat.

    .. attribute:: participants

        :class:`~whalesong.managers.group_metadata.ParticipantCollectionManager`

        Group's participants collection manager.
    """

    MODEL_CLASS = GroupMetadata

    def __init__(self, driver, manager_path=''):
        super(GroupMetadataManager, self).__init__(driver=driver, manager_path=manager_path)

        self.add_submanager('participants', ParticipantCollectionManager(
            driver=self._driver,
            manager_path=self._build_command('participants')
        ))

    def group_invite_code(self) -> Result[None]:
        return self._execute_command('groupInviteCode')

    def revoke_group_invite(self) -> Result[None]:
        return self._execute_command('revokeGroupInvite')


class GroupMetadataCollectionManager(BaseCollectionManager[GroupMetadataManager]):
    MODEL_MANAGER_CLASS = GroupMetadataManager
